### ISSUE TITLE
11

### ISSUE THREAD

**Issue by bryanjohnabraham** [bryanjohnabraham/ml#13]
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

# 1. Load Iris dataset
iris = load_iris()

# Use two features for 2D visualization
X = iris.data[:, [0, 1]]   # Sepal length, Sepal width
y = iris.target

# 2. Convert into binary classification
# Setosa = 0
# Non-Setosa = 1
y_binary = np.where(y == 0, 0, 1)

# 3. Split dataset
X_train, X_test, y_train, y_test = train_test_split(
    X, y_binary,
    test_size=0.2,
    random_state=42,
    stratify=y_binary
)

# 4. Standardize features
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# 5. Create and train Linear SVM
model = SVC(kernel='linear', C=1.0)

model.fit(X_train, y_train)

# 6. Prediction
y_pred = model.predict(X_test)

# 7. Accuracy
accuracy = accuracy_score(y_test, y_pred)

print("Linear SVM Accuracy:", accuracy)

# 8. Get weights and bias
w = model.coef_[0]
b = model.intercept_[0]

print("Weights:", w)
print("Bias:", b)

# 9. Calculate margin
margin = 2 / np.linalg.norm(w)

print("Margin width:", margin)

# 10. Plot data points
plt.figure(figsize=(8, 6))

plt.scatter(
    X_train[y_train == 0, 0],
    X_train[y_train == 0, 1],
    label="Setosa",
    marker='o'
)

plt.scatter(
    X_train[y_train == 1, 0],
    X_train[y_train == 1, 1],
    label="Non-Setosa",
    marker='s'
)

# 11. Create x values
x_values = np.linspace(
    X_train[:, 0].min() - 1,
    X_train[:, 0].max() + 1,
    200
)

# Decision boundary:
# w0*x + w1*y + b = 0
y_boundary = -(w[0] * x_values + b) / w[1]

# Margin boundaries:
# w0*x + w1*y + b = +1
y_margin_positive = -(w[0] * x_values + b - 1) / w[1]

# w0*x + w1*y + b = -1
y_margin_negative = -(w[0] * x_values + b + 1) / w[1]

# Plot decision boundary
plt.plot(
    x_values,
    y_boundary,
    label="Decision Boundary"
)

# Plot margins
plt.plot(
    x_values,
    y_margin_positive,
    '--',
    label="Margin"
)

plt.plot(
    x_values,
    y_margin_negative,
    '--'
)

# Highlight support vectors
plt.scatter(
    model.support_vectors_[:, 0],
    model.support_vectors_[:, 1],
    s=100,
    facecolors='none',
    edgecolors='black',
    label="Support Vectors"
)

plt.xlabel("Standardized Sepal Length")
plt.ylabel("Standardized Sepal Width")
plt.title("Linear SVM - Setosa vs Non-Setosa")
plt.legend()
plt.grid(True)

plt.show()
